src/service/service.py

```
@@ -451,6 +509,29 @@ def emit_progress(node_name: str | None) -> None:
+                credit_service = get_credit_service()
+
+                for feature in credit_usage["used_features"]:
+                    try:
+                        await credit_service.use_credit(
+                            user_id=user_id,
+                            feature_name=feature,
+                            difficulty=credit_usage.get("difficulty", "easy"),
+                            token=auth_token,
+                        )
+                        logger.info(f"Credit deducted for feature: {feature}")
+                    except Exception as credit_err:
```

Ruff (0.15.0)

[warning] Do not catch blind exception: `Exception`

(BLE001)